File: src/main/java/me/zeroeightsix/kami/module/modules/render/ArmourHUD.java
package me.zeroeightsix.kami.module.modules.render;

import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import me.zeroeightsix.kami.util.ColourHolder;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderItem;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;
import net.minecraft.world.GameType;

/**
 * Created by 086 on 24/01/2018.
 */
@Module.Info(name = "ArmourHUD", category = Module.Category.GUI, showOnArray = Module.ShowOnArray.OFF)
public class ArmourHUD extends Module {

    private static RenderItem itemRender = Minecraft.getMinecraft().getRenderItem();

    private Setting<Boolean> damage = register(Settings.b("Damage", false));

    private NonNullList<ItemStack> getArmour() {
        if (mc.playerController.getCurrentGameType().equals(GameType.CREATIVE) || mc.playerController.getCurrentGameType().equals(GameType.SPECTATOR)) {
            return NonNullList.withSize(4, ItemStack.EMPTY);
        }
        else {
            return mc.player.inventory.armorInventory;
        }
    }

    @Override
    public void onRender() {
        GlStateManager.enableTexture2D();

        ScaledResolution resolution = new ScaledResolution(mc);
        int i = resolution.getScaledWidth() / 2;
        int iteration = 0;
        int y = resolution.getScaledHeight() - 55 - (mc.player.isInWater() ? 10 : 0);

        for (ItemStack is : getArmour()) {
            iteration++;
            if (is.isEmpty()) continue;
            int x = i - 90 + (9 - iteration) * 20 + 2;
            GlStateManager.enableDepth();

            itemRender.zLevel = 200F;
            itemRender.renderItemAndEffectIntoGUI(is, x, y);
            itemRender.renderItemOverlayIntoGUI(mc.fontRenderer, is, x, y, "");
            itemRender.zLevel = 0F;

            GlStateManager.enableTexture2D();
            GlStateManager.disableLighting();
            GlStateManager.disableDepth();

            String s = is.getCount() > 1 ? is.getCount() + "" : "";
            mc.fontRenderer.drawStringWithShadow(s, x + 19 - 2 - mc.fontRenderer.getStringWidth(s), y + 9, 0xffffff);

            if (damage.getValue()) {
                float green = ((float) is.getMaxDamage() - (float) is.getItemDamage()) / (float) is.getMaxDamage();
                float red = 1 - green;
                int dmg = 100 - (int) (red * 100);
                mc.fontRenderer.drawStringWithShadow(dmg + "", x + 8 - mc.fontRenderer.getStringWidth(dmg + "") / 2, y - 11, ColourHolder.toHex((int) (red * 255), (int) (green * 255), 0));
            }
        }
    GlStateManager.enableDepth();
    GlStateManager.disableLighting();
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/render/ArmourHUD.kt
package me.zeroeightsix.kami.module.modules.render

import me.zeroeightsix.kami.event.events.RenderOverlayEvent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Settings
import me.zeroeightsix.kami.util.color.ColorConverter
import me.zeroeightsix.kami.util.event.listener
import net.minecraft.client.gui.ScaledResolution
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import net.minecraft.util.math.BlockPos
import net.minecraft.world.GameType
import kotlin.math.floor

@Module.Info(
        name = "ArmourHUD",
        category = Module.Category.RENDER,
        description = "Displays your armour and it's durability on screen",
        showOnArray = Module.ShowOnArray.OFF
)
object ArmourHUD : Module() {
    private val damage = register(Settings.b("Damage", false))
    private val armour: NonNullList<ItemStack>
        get() = if (mc.playerController.getCurrentGameType() == GameType.CREATIVE || mc.playerController.getCurrentGameType() == GameType.SPECTATOR) {
            NonNullList.withSize(4, ItemStack.EMPTY)
        } else {
            mc.player.inventory.armorInventory
        }

    init {
        listener<RenderOverlayEvent> {
            val resolution = ScaledResolution(mc)
            val width = resolution.scaledWidth / 2
            val height = resolution.scaledHeight - 55 - if (isEyeInWater()) 10 else 0

            for ((index, itemStack) in armour.withIndex()) {
                if (itemStack.isEmpty()) continue
                val x = width - (9 - index) * 20 + 92

                GlStateManager.enableDepth()
                GlStateManager.enableTexture2D()
                mc.renderItem.zLevel = 200f
                mc.renderItem.renderItemAndEffectIntoGUI(itemStack, x, height)
                mc.renderItem.renderItemOverlayIntoGUI(mc.fontRenderer, itemStack, x, height, "")
                mc.renderItem.zLevel = 0f
                GlStateManager.enableTexture2D()
                GlStateManager.disableLighting()
                GlStateManager.disableDepth()

                if (damage.value) {
                    val dura = (itemStack.maxDamage - itemStack.itemDamage) / itemStack.maxDamage.toFloat()
                    val duraText = dura.toInt().toString()
                    val green = (dura * 255.0f).toInt()
                    val red = 255 - green
                    mc.fontRenderer.drawStringWithShadow(duraText, x + 8 - mc.fontRenderer.getStringWidth(duraText) / 2.0f, height - 11.0f, ColorConverter.rgbToHex(red, green, 0))
                }
            }

            GlStateManager.enableDepth()
        }
    }

    private fun isEyeInWater(): Boolean {
        val eyePos = mc.player.getPositionEyes(1f)
        val flooredEyePos = BlockPos(floor(eyePos.x), floor(eyePos.y), floor(eyePos.z))
        val block = mc.world.getBlockState(flooredEyePos).block
        return block == Blocks.WATER || block == Blocks.FLOWING_WATER
    }
}